### ISSUE TITLE
javalib Datagram socket is not implemented.

### ISSUE THREAD

**Issue by LeeTibbert** [scala-native/scala-native#2279]
Percy Bysshe Shelley:
``` 
My name is Ozymandias, King of Kings;
Look on my Works, ye Mighty, and despair!
```

PlainSocketImpl.scala:
```
 override def create(streaming: Boolean): Unit = {
    val sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM, 0)
    if (sock < 0) throw new IOException("Couldn't create a socket")
    fd = new FileDescriptor(sock)
  }

```
Note that variable streaming is entirely ignored and the assumption is that
every socket is SOCK_STREAM.    

Same assumption is made in code in `bind()` & `connect()`.  The 
code in question in those two places is not the most direct 
and should be eliminated, not corrected.

**Comment by LeeTibbert** [comment]
I discovered this defect whilst trying to come up with a more elegant solution to
review comments for memory handling in  `bind()` & `connect()` PlainSocketImpl calls.

**Comment by LeeTibbert** [comment]
To be explicit, the concern is verification of socket type:  SOCK_DGRAM should be handled.
SOCK_SEQPACKET, SOCK_RAW, etc, should be rejected until support is provided.

Yes, it would be nice in the out years to support SOCK_SEQPACKET.

**Comment by LeeTibbert** [comment]
Turns out that Java 8 `Socket` class, at least, deprecates the constructors which specify stream.

The recommended DatagramSocket class appears to not be implemented in Scala Native.
Call Herakles, there is a 13th labor....

**Comment by LeeTibbert** [comment]
The reason that lack of Datagram may be a concern is that HTTP3 uses QUIC, which 
uses UDP (a.k.a Datagram).

**Comment by armanbilge** [comment]
FWIW I would be shocked if lack of UDP is the blocker for a QUIC implementation :)

For reference, here's a Java implementation of QUIC: https://github.com/ptrd/kwik

If we get to a place where we can be concerned that UDP is what prevents a Scala Native QUIC implementation, that would be a good day indeed!

---

Putting aside QUIC, where I do think UDP would be useful today is for implementing distributed tracing and metrics. For example AWS X-Ray is implemented with UDP.
https://docs.aws.amazon.com/xray/latest/devguide/xray-daemon.html

And here's a Scala X-Ray client that cross-compiles to JVM and JS, but not yet to Native due to lack of UDP support.
https://github.com/tpolecat/natchez/tree/series/0.1/modules/xray/shared/src/main/scala/natchez/xray

**Comment by LeeTibbert** [comment]
Thank you for the URLs.  All three are interesting.  

My main reason for updating this issue was to provide some background why it was
not just a crank "oh there is some incompleteness in javalib' Issue, which could fall
to the back of the pack.


> that would be a good day indeed! 
A good day to be sure.

I had a relatively good experience with porting the RE2 regex code from the java implementation
to Scala Native using a web site and _lots_ of hand touch up.  So the day of your amazement may
happen this decade or century.

**Comment by armanbilge** [comment]
I took a brief look at implementing Datagram in fs2-io directly in terms of Posix APIs, without using the JDK APIs.

https://github.com/typelevel/fs2/blob/51c976855db1afa73c3ff6f6f919aa131076d8c7/io/js-jvm/src/main/scala/fs2/io/net/DatagramSocket.scala

Unfortunately, the `DatagranSocket` API in FS2 still depends on `NetworkInterface`.

https://docs.oracle.com/javase/8/docs/api/java/net/NetworkInterface.html

**Comment by LeeTibbert** [comment]
Arman,

May not seem like it to you, but that is good info.  You have narrowed the scope of the problem. (That is an IPv6
in joke having to do with the "::1%en0" interface_id, also know as scope_id).

```
def join(
      join: MulticastJoin[IpAddress],
      interface: DatagramSocket.NetworkInterface
  ): F[GroupMembership]
```

`DataGramSocket.NetworkInterface` looks like fs2-io not Java. In a quick run through the fs2-io repository
I could find it in the DataGramSocket trait. It must come from something, which I also could not find, which
that trait extends.   I'll have to clone the Repository and rat around when I am rested.

It looks like this code works on Scala.js.  How does Scala.js provide the interface?  I took a quick look
there also and could not fine anything.    Like my local baseball team, now out of the run for the World Series,
I am striking out all over.

I whistled may way around tacking the SN NetworkInterface class when doing my recent IPv6
InetAddress "::1%en0" changes, but had the fear that my luck would not hold.   

The unix calls for the underlying information have existed for awhile. They are just detailed
and, from afar, messy (i.e. fear provoking).  As always, I have no idea of what happens on
Windows.

I am out-of-service for a few days, but after that, perhaps we should chat offline to see
if some subset might be useful to fs2-io.

**Comment by armanbilge** [comment]
Yes, `DatagramSocket.NetworkInterface` is an alias to help with cross-compiling.

---

On JVM it is a type alias for `java.net.NetworkInterface`.
https://github.com/typelevel/fs2/blob/51c976855db1afa73c3ff6f6f919aa131076d8c7/io/jvm/src/main/scala/fs2/io/net/DatagramSocketPlatform.scala#L40

This is to match the underlying JDK API.
https://docs.oracle.com/javase/8/docs/api/java/nio/channels/MulticastChannel.html#join-java.net.InetAddress-java.net.NetworkInterface-

---

On JS it is a type alias for `String`.
https://github.com/typelevel/fs2/blob/51c976855db1afa73c3ff6f6f919aa131076d8c7/io/js/src/main/scala/fs2/io/net/DatagramSocketPlatform.scala#L50

This is to match the underlying Node.js API.
https://nodejs.org/api/dgram.html#socketaddmembershipmulticastaddress-multicastinterface

---

The relevant Posix API appears to be the `IP_ADD_MEMBERSHIP` socket option. And analogously `IP_DROP_MEMBERSHIP`.
- https://man7.org/linux/man-pages/man7/ip.7.html
- https://www.freebsd.org/cgi/man.cgi?query=ip&sektion=4&manpath=FreeBSD+9.0-RELEASE

---

Another possibility I am seriously considering is adding a `NetworkInterface` datatype to the ip4s project, which already provides many useful datatypes that cross-compile for JVM, JS, and Native. Indeed, FS2 already uses the `SocketAddress` and `IpAddress` datatypes defined in that project.

https://github.com/comcast/ip4s

So far, the main reason I have hesitated is because on the JVM it appears to be impossible to create an instance of `NetworkInterface`; you can only acquire instances created by the JDK itself. While this does not seem unreasonable, it is not a great fit for the ip4s model which seeks to represent network data as pure Scala immutable classes, and convert directly to JDK APIs only when necessary for interop.

**Comment by LeeTibbert** [comment]
Arman, you have given me lots to study and several good starting points.

If I understand correctly, if SN implemented `java.net.NetworkInterface` that should work for fs2-io?
Mid week, I could do design study which started with a dummy, empty j.n.NetworkInterface class,
build fs2-io, and see what the linker reports as missing symbols.  Probably the static methods to
return a NetworkInterface instance. (getByAddress, getByMumble).  That would give me
a starting point.  

We agree that eventually, the full spec or as close to it as economic, needs to be implemented.
Sometime the Pareto 70/30 rule allows early work to earn its keep whilst more time consuming
work happens.

The idea, is for this to work starting in  the SN 0.4.0 stream, then 0.5.x? Reasonable, but to know
which ducks go first.

The posix socketopt routines will probably be necessary under the javalib covers to join IPv4 multicast groups.
IPv6 has slightly different concepts and definately  different constants.  Not needed for 0.4.x but
detailed work for 0.4.n.    

Sound like a plan?  It kicks implementing full DatagramSocket down the road a bit.

**Comment by armanbilge** [comment]
Yes, if SN implemented `java.net.NetworkInterface` I think that would be sufficient to unblock a `DatagramSocket` implementation based directly on Posix APIs (and also another implementation in fs2-io_uring).

Yes, we definitely need some static methods to obtain a `NetworkInterface` instance. And we would also need instance methods that would allow to create a `struct ip_mreqn` to use with the `IP_ADD_MEMBERSHIP`. I am hoping this is sufficient 😅

**Comment by LeeTibbert** [comment]
That information bounds the problem.  I'll start a private design study and see how it goes. 
Should be a good learning exercise.   

I am out a couple of days this week, so progress will not be immediate, but should happen.

**Comment by LeeTibbert** [comment]
I have the Java-to-Scala parts of `NetworkInterface` & the `InterfaceAddress` class it may need mocked out.
Now to implement methods in detail.

>  instance methods that would allow to create a struct ip_mreqn to use with the IP_ADD_MEMBERSHIP socket option.
`struct ip_mreqn` looks like a posixlib structure for `setsockopt`.  I took a quick look at the structure and it looks
like it could be filled in.

My next step is to clone the fs2-io code and see how this stuff is actually called, in Java.  Implementing full
classes is the ultimate goal, but knowing the immediate calls will help me focus my efforts. When I can return
to this (a few days), I will chase the URLs you posted above.  Thanks.
